lib/amplitude/event.rb

```diff
@@ -0,0 +1,285 @@
+      @retry = 0
+    end
+
+    def [](key)
+      instance_variable_get("@#{key}")
+    end
+
+    def []=(key, value)
+      send("#{key}=", value) if verify_property(key, value)
+    end
+
+    def include?(key)
+      instance_variable_defined?("@#{key}") && !instance_variable_get("@#{key}").nil?
+    end
+
+    def valid_properties?(key, value)
+      return false unless key.is_a?(String)
+
+      if value.is_a?(Array)
+        result = true
+        value.each do |element|
+          return false if element.is_a?(Array)
+
+          if element.is_a?(Hash)
+            result &&= valid_object?(element)
+          elsif !element.is_a?(Numeric) && !element.is_a?(String) && !element.is_a?(TrueClass) && !element.is_a?(FalseClass)
+            result = false
+          end
+          break unless result
+        end
+        return result
+      end
+
+      return valid_object?(value) if value.is_a?(Hash)
+
+      value.is_a?(TrueClass) || value.is_a?(FalseClass) ||
+        value.is_a?(Numeric) || value.is_a?(String) || value.is_a?(Symbol)
+    end
+
+    def valid_object?(obj)
+      obj.each do |key, value|
+        return false unless valid_properties?(key, value)
+      end
+      true
+    end
+
+    def verify_property(key, value)
+      return true if value.nil?
+
+      unless instance_variable_defined?("@#{key}")
+        logger.error("Unexpected event property key: #{key}")
+        return false
+      end
+
+      expected_type = EVENT_KEY_MAPPING[key][1]
+      unless value.is_a?(expected_type)
+        logger.error("Event property #{key} expected #{expected_type} but received #{value.class}.")
+        return false
+      end
+
+      return valid_object?(value) if value.is_a?(Hash)
+
+      true
+    end
+
+    def event_body
+      event_body = {}
+      EVENT_KEY_MAPPING.each do |key, mapping|
+        next unless include?(key) && self[key]
+
+        value = self[key]
+        if value.is_a?(mapping[1])
+          event_body[mapping[0]] = value
+        else
+          puts "#{self.class}.#{key} expected #{mapping[1]} but received #{value.class}."
+        end
+      end
+      event_body['plan'] = @plan.plan_body if @plan.respond_to?(:plan_body)
+      event_body['ingestion_metadata'] = @ingestion_metadata.body if @ingestion_metadata.respond_to?(:body)
+      %w[user_properties event_properties group_properties].each do |properties|
+        next unless event_body[properties]
+        # TODO: check ENUM
+        # event_body[properties].transform_values! do |value|
+        # value.is_a?(Enum) ? value.value : value
+        # end
+      end
+      AmplitudeAnalytics.truncate(event_body.sort.to_h)
+    end
+
+    def callback(status_code, message = nil)
+      @event_callback.call(self, status_code, message) if @event_callback.respond_to?(:call)
+    end
+
+    def to_s
+      JSON.generate(event_body)
+    end
+  end
+
+  # BaseEvent
+  class BaseEvent < EventOptions
+    def initialize(event_type, **kwargs)
```

Comment:
Have limited knowledge about Ruby. Afaik, Ruby can be used for both client-side and server-side programming, but it's primarily known for server-side development. 
- if it's a server-side SDK, `BaseEvent` should require `event_type` and either `user_id` or `device_id`
- if it's a client-side SDK, device Id is generated when SDK init

Reply:
Thanks for the detailed comments! The check for `user_id` or `device_id` has been fixed